core/src/main/scala/tapir/internal/server/DecodeInputs.scala

```diff
@@ -71,119 +73,204 @@ object DecodeInputs {
 
     // we're using null as a placeholder for the future values. All except the body (which is determined by
     // interpreter-specific code), should be filled by the end of this method.
     compose(
-      apply(methodInputs, _, _),
-      apply(pathInputs, _, _),
-      (values, ctx) =>
-        verifyPathExactMatch(pathInputs, ctx) match {
-          case None          => (values, ctx)
-          case Some(failure) => (failure, ctx)
-        },
-      apply(otherInputs, _, _)
+      matchOthers(methodInputs, _, _),
+      matchPath(pathInputs, _, _),
+      matchOthers(otherInputs, _, _)
     )(DecodeInputsResult.Values(Vector.fill(basicInputs.size)(null), None), ctx)._1
   }
 
-  private def apply(
-      inputs: Vector[(EndpointInput.Basic[_], Int)],
-      values: DecodeInputsResult.Values,
+  private def matchPath(
+      pathInputs: Vector[IndexedBasicInput],
+      decodeValues: DecodeInputsResult.Values,
       ctx: DecodeInputsContext
   ): (DecodeInputsResult, DecodeInputsContext) = {
-    inputs match {
-      case Vector() => (values, ctx)
+    pathInputs match {
+      case Vector() =>
+        // Match everything if no path input is specified
+        decodeValues -> ctx
+      case _ +: _ =>
+        matchAndDecodePath(pathInputs, ctx, decodeValues)
+    }
 
-      case (input @ EndpointInput.FixedMethod(m), _) +: inputsTail =>
-        if (m == ctx.method) apply(inputsTail, values, ctx)
-        else (DecodeInputsResult.Failure(input, DecodeResult.Mismatch(m.m, ctx.method.m)), ctx)
+  }
 
-      case (input @ EndpointInput.FixedPath(ss), _) +: inputsTail =>
-        ctx.nextPathSegment match {
-          case (Some(`ss`), ctx2)       => apply(inputsTail, values, ctx2)
-          case (None, ctx2) if ss == "" => apply(inputsTail, values, ctx2) // root path
-          case (Some(s), _)             => (DecodeInputsResult.Failure(input, DecodeResult.Mismatch(ss, s)), ctx)
-          case (None, _)                => (DecodeInputsResult.Failure(input, DecodeResult.Missing), ctx)
-        }
+  private def matchAndDecodePath(
+      nonEmptyInputs: Vector[IndexedBasicInput],
+      origCtx: DecodeInputsContext,
+      origDecodeValues: DecodeInputsResult.Values
+  ): (DecodeInputsResult, DecodeInputsContext) = {
 
-      case (input @ EndpointInput.PathCapture(codec, _, _), index) +: inputsTail =>
-        ctx.nextPathSegment match {
-          case (Some(s), ctx2) =>
-            codec.safeDecode(s) match {
-              case DecodeResult.Value(v)  => apply(inputsTail, values.setBasicInputValue(v, index), ctx2)
-              case failure: DecodeFailure => (DecodeInputsResult.Failure(input, failure), ctx)
+    var currCtx = origCtx
+    var decodeInputResult: DecodeInputsResult = origDecodeValues
+    var thunks = Vector.empty[() => (IndexedBasicInput, DecodeResult[_])]
+    nonEmptyInputs.foreach {
```

Comment:
So the crucial difference in this approach is that you first map all path segments to their decoding results (not checking if this is a successful or failed decoding at this stage). This is collected in the `thunks` variable.

Once this is done, you then check if there are remaining path segments. If yes - the decoding fails with a `Mismatch` and this **takes priority** over any potential failures in decoding the segments.

Then you take into account the values returned by decoding the segments.

This is all fine, however I would try to recfactor this code a little:
1. break out the `foreach`, returning a vector of thunks, into a separate method
2. in the separate mehod, instead of using variables, try to use recursion (it's easier to read this way)
3. factor out the `recur` method into a properly named, top-level method

This should make this code easier to read. Would you be able to do that?

Reply:
Yep that's an accurate summary of the logic.

I started with recursion (and got it working), but there's a problem that made the code require mutual recursion (i.e. no tailrec).

The issue is [here](https://github.com/softwaremill/tapir/pull/160/files/7d33485adb217f871b3e860d8b0952c67297689a#diff-7886d7396a1a7018e63d215193279df9R186).
When there's an extra segment found in the request URI (e.g. wanted: `/customer/*` got: `/customer/*/blah`). You need to return DecodeInputsResult.Failure which require a `EncodeInput.Basic[_]` - by using recursion, you have no guarantee that the passed in `Vector[IndexedBasicInput]` is nonempty and thus cannot reliably provide a `EncodeInput.Basic[_]` for the Failure constructor. 
There are ways to work around it (by passing a backup `EncodeInput.Basic[_]` to blame on?), but it makes the recursion function interface more confusing to understand.

Will address the other comments and add more comments.

Reply:
ah ... ok, well, let's see where you'll get to :)

Reply:
Done. Passing in a fallback input to blame a failure turned out ok :)